### PR TITLE
joining an unloaded room

### LINKED ISSUE #24: specify what happens when you try to join a permanent room that is not loaded

### PR DESCRIPTION
**Pull request by sophiadicuffa**
- specify details about joining an unloaded permanent room
- add a sequence diagram for  joining an unloaded permanent room case 
- added API proxy description

closes #24

### CODE REVIEW

File: final-project/manApiProxy.tex [hunk 1]
```diff
@@ -5,4 +5,11 @@ \chapter{REST API Proxy \\
 
 
 TODO: describe how REST API requests will be forwarded from the load balancers to the OTT monoliths
+
+When a REST API request is made to the OpenTogetherTube system, it is first received by the NGINX reverse proxy server. 
+NGINX then forwards the request to one of the OTT monoliths based on the current load balancing algorithm in use. 
+The monolith that receives the request will then process it and return a response, which will be sent back to the 
+NGINX reverse proxy server. The reverse proxy server will then forward the response back to the client that made the 
+original request.
```
Comment (dyc3) [inline final-project/manApiProxy.tex]: nit: nginx is usually all lowercase.

It's not necessarily nginx that is the reverse proxy. It can also be heroku or cloudflare, or in the current official site configuration cloudflare followed by heroku. These services are effectively stateless, so we can assume they are transparent, as in any request to any reverse proxy yields the same result.

Right now, if we take into consideration all of the components involved in serving a page from opentogethertube.com, you would observe the following. Note that this is not the only possible configuration (people self-host OTT).

```mermaid
flowchart LR
    Cloudflare --> Heroku --> OTT
```
With the smart load balancer it would look like this:
```mermaid
flowchart LR
    Cloudflare --> Heroku --> Balancer

    A[Monolith]
    B[Monolith]
    C[Monolith]

    Balancer --> A
    Balancer --> B
    Balancer --> C
```

So, you need to specify how the balancer is going to route requests to monoliths.

File: final-project/Figures/unloaded-room.png
Comment (dyc3) [inline final-project/Figures/unloaded-room.png:1]: This diagram does not account for the Balancer. 

nit: Also, the "redirect" terminology implies an HTTP redirect, which is _not quite_ what happens currently. So that should be clarified.

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: final-project/Figures/unloaded-room.png
Comment (sophiadicuffa) [reply final-project/Figures/unloaded-room.png:1]: would this be better?
```mermaid
sequenceDiagram
    actor User
    participant Load Balancer
    participant Server1
    participant Server2
    participant Database
    
    User->>+Load Balancer: Join Permanent Room
    Load Balancer->>+Server1: Forward Request
    Server1->>+Database: Check if Room is Loaded
    alt Room is Loaded
        Server1-->>-Load Balancer: Redirect to Room
    else Room is Not Loaded
        Database-->>Server1: Load Room
        Server1->>+Server1: Create New Room Instance
        Server1->>+Server1: Add to Loaded Rooms
        Server1-->>-Load Balancer: Redirect to Room
    end
    Load Balancer->>+Server2: Forward Request
    Server2->>+Database: Check if Room is Loaded
    alt Room is Loaded
        Server2-->>-Load Balancer: Redirect to Room
    else Room is Not Loaded
        Database-->>Server2: Load Room
        Server2->>+Server2: Create New Room Instance
        Server2->>+Server2: Add to Loaded Rooms
        Server2-->>-Load Balancer: Redirect to Room
    end
```

Comment (dyc3) [reply final-project/Figures/unloaded-room.png:1]: Not quite. 

- You should favor the established terminology "Balancer" and "Monolith" rather than "server1" and "server2". 
- The balancer the the entity that checks if the room is loaded.
- In this diagram, the user never receives any information about the room.

File: final-project/manApiProxy.tex [hunk 2]
```diff
@@ -3,6 +3,24 @@ \chapter{REST API Proxy \\
...
+then returns the response to the load balancer, which in turn sends the response back to the client that made the original
+request.
 
-TODO: describe how REST API requests will be forwarded from the load balancers to the OTT monoliths
-TODO: be sure to talk about how request credentials will be forwarded
\ No newline at end of file
+\begin{figure}[!htb]
+  \centering
+  \scalebox{0.57}{\includegraphics{Figures/api-balancer.png}}
+  \caption{\label{Figure::api-balancer} Unloaded Room Sequence Diagram.}
```
Comment (dyc3) [inline final-project/manApiProxy.tex]: incorrect caption

File: final-project/manApiProxy.tex [hunk 3]
```diff
@@ -3,6 +3,24 @@ \chapter{REST API Proxy \\
   \index{Chapter!rest-api-proxy}
   \label{Chapter::RestApiProxy}}
 
+\section{Forwarding API Requests}
+When a REST API request is made to OpenTogetherTube, the request is first received by the load balancer. The 
+load balancer then selects one of the OTT monoliths based on the current load balancing algorithm, and forwards the 
+request to the selected monolith's nginx reverse proxy server. The nginx reverse proxy server receives the request and
```
Comment (dyc3) [inline final-project/manApiProxy.tex]: > selected monolith's nginx reverse proxy server.

again, there is no nginx here. The balancer will act as the reverse proxy for API requests.

File: final-project/manApiProxy.tex [hunk 4]
```diff
@@ -3,6 +3,24 @@ \chapter{REST API Proxy \\
...
\ No newline at end of file
+\begin{figure}[!htb]
+  \centering
+  \scalebox{0.57}{\includegraphics{Figures/api-balancer.png}}
+  \caption{\label{Figure::api-balancer} Unloaded Room Sequence Diagram.}
+\end{figure}
+
+\section{Forwarding Requests Credentials}
+
+OpenTogetherTube platform uses JSON Web Tokens to authenticate and authorize users. When a user logs in, the
```
Comment (dyc3) [inline final-project/manApiProxy.tex]: OTT does not currently use JWTs. It just generates and issues "tokens" which are used to look up session info in redis.

We need to more thoroughly evaluate whether or not we should use JWTs, but that can be left for another PR.

File: final-project/manJoinProcess.tex [hunk 5]
```diff
@@ -64,3 +64,20 @@ \subsection{Messaging Protocol}
 \section{Creating or Loading Rooms}
 
 \subsection{Monolith Node Selection}
+
+\subsection{Unloaded Room}
+When a user attempts to join a permanent room that is not loaded in OpenTogetherTube, the server undergoes a series of
```
Comment (dyc3) [inline final-project/manJoinProcess.tex]: You need to specify if this is explaining how it currently is, or what it will look like with the load balancer.

File: final-project/manJoinProcess.tex [hunk 6]
```diff
@@ -64,3 +64,20 @@ \subsection{Messaging Protocol}
...
+ events. Firstly, the server receives a request from the user to join a permanent room. Secondly, it checks if the room 
+ is currently loaded in memory, and if it is, the user is directed to the room. However, if the room is not loaded, the 
+ server loads the room from the database and creates a new instance of the room. The new room instance is then added to a 
+ collection of loaded rooms by the server. Once the room has been successfully loaded, the user is directed to the room.
+
+ \begin{figure}[!htb]
+  \centering
+  \scalebox{0.57}{\includegraphics{Figures/unloaded-room.png}}
+  \caption{\label{Figure::unloaded-room} Unloaded Room Sequence Diagram.}
```
Comment (dyc3) [inline final-project/manJoinProcess.tex:80]: Elaborate: is this for OTT's current architecture, or the proposed new one?

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: final-project/manJoinProcess.tex (hunk 6 above)
Comment (sophiadicuffa) [reply final-project/manJoinProcess.tex:78]: im going off the diagram you put in the comment so I'm not sure

Comment (sophiadicuffa) [reply final-project/manJoinProcess.tex:80]: i thought the new one, what is wrong with it?

Comment (dyc3) [reply final-project/manJoinProcess.tex:80]: The diagram looks like it describes the current flow. It doesn't account for the load balancer, and it doesn't use the established terminology to refer to the balancer and monolith.

This is fine, as long as you make that clear in the caption. If that's not what you intended, you need to fix the diagram.

File: final-project/manApiProxy.tex [hunk 7]
```diff
@@ -4,23 +4,22 @@ \chapter{REST API Proxy \\
...
 \section{Forwarding API Requests}
-When a REST API request is made to OpenTogetherTube, the request is first received by the load balancer. The 
-load balancer then selects one of the OTT monoliths based on the current load balancing algorithm, and forwards the 
-request to the selected monolith's nginx reverse proxy server. The nginx reverse proxy server receives the request and 
-forwards it to the corresponding monolith for processing. 
-The monolith processes the request and sends a response back to its nginx reverse proxy server. The reverse proxy server 
-then returns the response to the load balancer, which in turn sends the response back to the client that made the original
-request.
+When a REST API request is made to OpenTogetherTube, the request is first received by the load balancer
+ acting as the reverse proxy for API requests. The load balancer then selects one of the OTT monoliths based
```
Comment (dyc3) [inline final-project/manApiProxy.tex:8]: > The load balancer then selects one of the OTT monoliths based on the current load balancing algorithm and forwards the request to the selected monolith.

This is not true for all requests. For example, all endpoints that start with `/api/room/<roomname>` need to be routed to a specific monolith if the room is loaded.

I'm fine with merging this as is, as long as you open an issue saying this needs to be fixed.

File: final-project/manApiProxy.tex [hunk 8]
```diff
@@ -4,23 +4,22 @@ \chapter{REST API Proxy \\
...
 \section{Forwarding Requests Credentials}
 
-OpenTogetherTube platform uses JSON Web Tokens to authenticate and authorize users. When a user logs in, the 
-server generates a JWT that is stored in the browser's local storage and included in all subsequent requests made
-by the user. The server verifies the JWT to ensure that the user is authenticated and has permission to access the 
-requested video stream.
\ No newline at end of file
+OpenTogetherTube platform uses tokens to authenticate and authorize users. When a user logs in, the
+server generates a token that is stored in the browser's local storage and included in all subsequent requests made
+by the user. The server verifies the token to ensure that the user is authenticated and has permission to access the
```
Comment (dyc3) [inline final-project/manApiProxy.tex:23]: > The server verifies the token to ensure that the user is authenticated and has permission to access the
requested video stream.

OTT does not stream video. The video stream comes from external services. The token is there to identify the user. Otherwise, this paragraph is correct.

File: final-project/manJoinProcess.tex [hunk 9]
```diff
@@ -66,11 +66,13 @@ \section{Creating or Loading Rooms}
 \subsection{Monolith Node Selection}
 
 \subsection{Unloaded Room}
-When a user attempts to join a permanent room that is not loaded in OpenTogetherTube, the server undergoes a series of
- events. Firstly, the server receives a request from the user to join a permanent room. Secondly, it checks if the room 
- is currently loaded in memory, and if it is, the user is directed to the room. However, if the room is not loaded, the 
- server loads the room from the database and creates a new instance of the room. The new room instance is then added to a 
- collection of loaded rooms by the server. Once the room has been successfully loaded, the user is directed to the room.
+When a user tries to join a permanent room in OpenTogetherTube, the request is first received by the load balancer. 
+The load balancer forwards the request to one of the available servers (Server1 or Server2) based on the current load
```
Comment (dyc3) [inline final-project/manJoinProcess.tex:70]: need to use "Monolith" terminology rather than "Server1" and "Server2"

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: final-project/manApiProxy.tex (hunk 7 above)
Comment (sophiadicuffa) [reply final-project/manApiProxy.tex:8]: issued opened. #32

File: final-project/Figures/unloaded-room.png
Comment (sophiadicuffa) [reply final-project/Figures/unloaded-room.png:1]: here is an updated version
```mermaid
sequenceDiagram
    actor User
    participant Balancer
    participant Monolith1 as Monolith1
    participant Monolith2 as Monolith2
    participant Database
    
    User->>+Balancer: Join Permanent Room
    Balancer->>+Monolith1: Forward Request
    Monolith1->>+Database: Check if Room is Loaded
    alt Room is Loaded
        Monolith1-->>-Balancer: Redirect to Room
        Balancer-->>User: Send Room Information
    else Room is Not Loaded
        Database-->>Monolith1: Load Room
        Monolith1->>+Monolith1: Create New Room Instance
        Monolith1->>+Monolith1: Add to Loaded Rooms
        Monolith1-->>-Balancer: Redirect to Room
        Balancer-->>User: Send Room Information
    end
    Balancer->>+Monolith2: Forward Request
    Monolith2->>+Database: Check if Room is Loaded
    alt Room is Loaded
        Monolith2-->>-Balancer: Redirect to Room
        Balancer-->>User: Send Room Information
    else Room is Not Loaded
        Database-->>Monolith2: Load Room
        Monolith2->>+Monolith2: Create New Room Instance
        Monolith2->>+Monolith2: Add to Loaded Rooms
        Monolith2-->>-Balancer: Redirect to Room
        Balancer-->>User: Send Room Information
    end
```

Comment (dyc3) [reply final-project/Figures/unloaded-room.png:1]: This is fine for now.

Comment (dyc3) [reply final-project/Figures/unloaded-room.png:1]: This is fine for now, I've opened #33 to return to this later.

**Review by dyc3 — APPROVED** [review]
